### PR TITLE
Added Shoot Effect Section and removed Shoot effect

### PR DESCRIPTION
**Pull request by TFSMads**
### Description
I have added a effect section to the shoot effect and changed LastSpawnedEntity effect to use new effect section instead of the previous effect class.

---
**Target Minecraft Versions:** 1.11 and newer
**Requirements:** No additional requirements
**Related Issues:** https://github.com/SkriptLang/Skript/issues/4789

### CODE REVIEW

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:53]:
```suggestion
	"You can modify the entity in this section, using for example 'event-entity' or 'event-arrow'.",
```

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:63]:
```suggestion
	"# Shoot section example",
	"command /firearrow:",
	"\ttrigger:",
	"\t\tshoot an arrow:",
	"\t\t\tignite event-projectile"})
```

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:89]: You should directly reference `ShootEvent.class` rather than `EffSecShoot.ShootEvent.class`

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Looks really good overall. Really just some formatting changes :)

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:97]: You can remove this check as it was for 1.9 but we no longer support that version

Comment (TFSMads) [reply src/main/java/ch/njol/skript/sections/EffSecShoot.java:97]: The org.bukkit.util.Consumer Class wasn't added berfore 1.11 so it isn't availble in 1.10 either.

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:28]: You should only import the specific classes needed :)

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:205]: Here too

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:206]: Split this line please, ~120 chars per line.

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:171]: Is `final` modifier here needed?

File: src/test/skript/tests/syntaxes/sections/EffSecShoot.sk
Comment (AyhamAl-Ali) [inline src/test/skript/tests/syntaxes/sections/EffSecShoot.sk:3]:
```suggestion
	shoot an arrow from spawn of "world":
```

Comment (AyhamAl-Ali) [reply src/test/skript/tests/syntaxes/sections/EffSecShoot.sk:3]: Replace it in other places too

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (TFSMads) [reply src/main/java/ch/njol/skript/sections/EffSecShoot.java:171]: No it isn't i just missed it when removing the others.

Comment (AyhamAl-Ali) [reply src/main/java/ch/njol/skript/sections/EffSecShoot.java]: No worries \:)

**Review by TheLimeGlass — APPROVED** [review]

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:86]:
```suggestion
		}

```

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:101]:
```suggestion
	private static final Double DEFAULT_SPEED = 5.0;
```

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:160]:
```suggestion
		if (v != null) {
```

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:162]:
```suggestion
			if (dir != null) {
```

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:174]: Shouldn't be needed since you suppress this for the whole method

We also don't use suppressions for individual lines

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
Formatting changes

Overall looking pretty good!! Thanks for this :)

File: src/test/skript/tests/syntaxes/sections/EffSecShoot.sk
Comment (AyhamAl-Ali) [inline src/test/skript/tests/syntaxes/sections/EffSecShoot.sk:4]:
```suggestion
		assert {_foo} is true with "variables from previous section wasn't carried over"
```

File: src/main/java/ch/njol/skript/expressions/ExprLastSpawnedEntity.java
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprLastSpawnedEntity.java:81]: Since you're editing it \:)
```suggestion
		final Entity en = from == 0 ? EffSecSpawn.lastSpawned : from == 1 ? EffSecShoot.lastSpawned : EffDrop.lastSpawned;
```

**Review by AyhamAl-Ali — CHANGES_REQUESTED** [review]
Pretty good so far \:)

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:100]:
```suggestion
		}, EventValues.TIME_NOW);
```

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:155]:
```suggestion
				trigger.execute(shootEvent);
				Variables.setLocalVariables(e, localVars); // Carry over variable changes to the rest of the main trigger
```
This is something that needs addressed in these "spawn" type sections.
Basically, variable changes made within a Section like this should be reflected for the rest of the trigger outside of the section.

I would make sure to add some testing for this in your test script.

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:177]: This method is nullable so you should check for null just in case

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:172]: This method is nullable so you should check for null just in case

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

**Comment by TPGamesNL** [comment]
This should have the same improvements and fixes applied to it as https://github.com/SkriptLang/Skript/pull/4911 (i.e. not allow delays in the section and copy local variables back). I recommend you wait for that PR to be merged / get more approval, then you know how to implement it here.

**Comment by TFSMads** [comment]
The two commits is the same, i accidentally commited twice.

**Comment by TFSMads** [comment]
> This should have the same improvements and fixes applied to it as #4911 (i.e. not allow delays in the section and copy local variables back). I recommend you wait for that PR to be merged / get more approval, then you know how to implement it here.

Can you review the changes i made in commit https://github.com/SkriptLang/Skript/pull/4792/commits/e3b9a5e04cc3c6d64c0314a6642e869f81846922.

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:159]: You don't need to manually remove them, it will be done automatically iirc

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
I think you should use TP's suggestion instead of adding/changing methods in Section/Trigger

**Comment by Moderocky** [comment]
Hello, we're closing this PR as it has changes requested and no activity for over six months.

You may re-open this whenever you're ready to come back to it.
In future, you can mark PRs as draft if you do not plan to work on them.

**Pull request closed without merging**
